src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/DefaultGetRecordsCache.java

```diff
@@ -1,48 +1,134 @@
+            this.maxSize = maxSize;
+            this.maxByteSize = maxByteSize;
+            this.maxRecordsCount = maxRecordsCount;
+            this.getRecordsResultQueue = new LinkedBlockingQueue<>(this.maxSize);
+        }
+    }
+    
+    private void start() {
+        if (dataFetchingStrategy.equals(DataFetchingStrategy.PREFETCH_CACHED)) {
+            log.info("Starting prefetching thread.");
+            executorService.execute(new DefaultGetRecordsCacheDaemon());
+        }
+        started = true;
     }
 
     @Override
     public GetRecordsResult getNextResult() {
-        return null;
+        if (!started) {
+            start();
+        }
+        GetRecordsResult result = null;
+        if (dataFetchingStrategy.equals(DataFetchingStrategy.PREFETCH_CACHED)) {
+            try {
+                result = getRecordsResultQueue.take();
+                updateBytes(result, false);
+                updateRecordsCount(result, false);
+            } catch (InterruptedException e) {
+                log.error("Interrupted while getting records from the cache", e);
+            }
+        } else {
+            result = validateGetRecordsResult(getRecordsRetrievalStrategy.getRecords(maxRecordsPerCall));
+        }
+        return result;
     }
 
     @Override
     public void shutdown() {
-
+        executorService.shutdown();
     }
+    
+    private void updateBytes(final GetRecordsResult getRecordsResult, final boolean add) {
+        getRecordsResult.getRecords().forEach(record -> {
+            int newLength = record.getData().array().length;
+            if (add) {
+                currentSizeInBytes += newLength;
+            } else {
+                currentSizeInBytes -= newLength;
+            }
+        });
+    }
+    
+    private void updateRecordsCount(final GetRecordsResult getRecordsResult, final boolean add) {
+        int newSize = getRecordsResult.getRecords().size();
+        if (add) {
+            currentRecordsCount += newSize;
+        } else {
+            currentRecordsCount -= newSize;
+        }
+    }
+    
+    private GetRecordsResult validateGetRecordsResult(final GetRecordsResult getRecordsResult) {
+        if (getRecordsResult == null) {
+            return new GetRecordsResult().withRecords(Collections.emptyList());
+        }
+        return getRecordsResult;
+    }
+    
+    private class DefaultGetRecordsCacheDaemon implements Runnable {
+        @Override
+        public void run() {
+            while (true) {
+                if (currentSizeInBytes < maxByteSize && currentRecordsCount < maxRecordsCount) {
+                    try {
+                        GetRecordsResult getRecordsResult = validateGetRecordsResult(
+                                getRecordsRetrievalStrategy.getRecords(maxRecordsPerCall));
+                        getRecordsResultQueue.put(getRecordsResult);
+                        if (getRecordsResultQueue.contains(getRecordsResult)) {
```

Comment:
Put can't fail, and contains is a linear operation.  This looks like an attempt to deal with the a consumer immediately pulling the value off the queue.  With the way we're managing the values this may not be a problem.   It does seem the intent is to mutate things synchronously though.

Reply:
This as we talked was being done, because of getRecords from KinesisDataFetcher being able to return nulls back. The new commit to this PR includes changes that makes sure that the getRecords doesn't return nulls anymore.